PyBuilder improperly handles syntactical errors in test being loaded

Details here:

https://gitter.im/pybuilder/pybuilder?at=6077844f81866c680c23d17d

Logs here:

https://files.gitter.im/5a174492d73408ce4f800535/d6Uj/build_logs.txt

I believe this is the same case I was seeing, with the symptom being:

```
BUILD FAILED - TypeError: 'NoneType' object is not callable (xmlrunner/runner.py:66)
```

For the people coming here from search, `-vX -Premote_debug=2` parameters to `pyb` will unearth the actual issue, which in my case was:

```
[DEBUG/MainProcess] received (0, TypeError('__init__() takes 1 positional argument but 2 were given'))
```

--- a/src/integrationtest/python/issue_770_tests.py
+++ b/src/integrationtest/python/issue_770_tests.py
@@ -1,0 +1,57 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2021 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+import unittest
+
+from itest_support import IntegrationTestSupport
+
+
+class Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.unittest")
+
+default_task = ["verify"]
+
+@init
+def init (project):
+    project.set_property("remote_debug", 2)
+""")
+        self.create_directory("src/main/python")
+        self.create_directory("src/unittest/python")
+        self.write_file("src/unittest/python/issue_770_tests.py", r"""
+print("Hello")
+   print("World!")
+""")
+        reactor = self.prepare_reactor()
+        with self.assertRaises(IndentationError) as e:
+            reactor.build()
+
+        self.assertEqual(e.exception.lineno, 3)
+
+        # This doesn't work on PyPy 3.7
+        # self.assertEqual(e.exception.offset, 3)
+
+        self.assertEqual(e.exception.text, '   print("World!")\n')
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/main/python/pybuilder/plugins/python/remote_tools/unittest_tool.py
+++ b/src/main/python/pybuilder/plugins/python/remote_tools/unittest_tool.py
@@ -17,6 +17,7 @@
 #   limitations under the License.
 
 import sys
+
 from pybuilder.plugins.python.remote_tools import start_tool, RemoteObjectPipe, Tool, logger, PipeShutdownError
 
 __all__ = ["start_unittest_tool", "PipeShutdownError", logger]
@@ -37,9 +38,17 @@ class UnitTestTool(Tool):
         loader = unittest.defaultTestLoader
         if self.test_method_prefix:
             loader.testMethodPrefix = self.test_method_prefix
-        tests = loader.loadTestsFromNames(self.test_modules)
 
-        pipe.expose("unittest_tests", tests)
+        try:
+            tests = loader.loadTestsFromNames(self.test_modules)
+            pipe.expose("unittest_tests", tests)
+        except SystemExit as e:
+            raise e
+        except KeyboardInterrupt as e:
+            raise e
+        except Exception as e:
+            pipe.expose("unittest_tests", e, error=True)
+
         pipe.register_remote_type(unittest.BaseTestSuite)
         pipe.register_remote_type(unittest.TestCase)
 

--- a/src/main/python/pybuilder/remote/__init__.py
+++ b/src/main/python/pybuilder/remote/__init__.py
@@ -398,7 +398,7 @@ def proxy_members(obj, public=True, protected=True, add_callable=True, add_str=T
 
 class RemoteObjectPipe:
 
-    def expose(self, name, obj, methods=None, fields=None, remote=True):
+    def expose(self, name, obj, methods=None, fields=None, remote=True, error=False):
         """Same as `RemoteObjectManager.expose`"""
         raise NotImplementedError
 
@@ -499,13 +499,18 @@ class _RemoteObjectSession:
 
         return _RemoteObjectPipe(self)
 
-    def expose(self, name, obj, remote=True, methods=None, fields=None):
+    def expose(self, name, obj, remote=True, methods=None, fields=None, error=False):
         exposed_objs = self._exposed_objs
 
         if name in exposed_objs:
             raise ValueError("%r is already exposed" % name)
+
+        if error:
+            obj = ExposedObjectError(obj)
+
         exposed_objs[name] = obj
-        if remote:
+
+        if not error and remote:
             self.register_remote(obj, methods, fields)
 
     def hide(self, name):
@@ -602,6 +607,11 @@ class PipeShutdownError(RemoteObjectError):
         self.cause = cause
 
 
+class ExposedObjectError(RemoteObjectError):
+    def __init__(self, cause=None):
+        self.cause = cause
+
+
 class _BaseProxy:
     def __init__(self, __rop, __proxy_def):
         self.__rop = __rop
@@ -619,12 +629,13 @@ ROP_PICKLE_VERSION = 2
 
 ROP_GET_EXPOSED = 5
 ROP_GET_EXPOSED_RESULT = 6
+ROP_GET_EXPOSED_ERROR = 7
 
-ROP_GET_PROXY_DEF = 7
-ROP_GET_PROXY_DEF_RESULT = 8
+ROP_GET_PROXY_DEF = 10
+ROP_GET_PROXY_DEF_RESULT = 11
 
-ROP_VERIFY_TYPES = 9
-ROP_VERIFY_TYPES_RESULT = 10
+ROP_VERIFY_TYPES = 14
+ROP_VERIFY_TYPES_RESULT = 15
 
 ROP_REMOTE_ACTION = 20
 ROP_REMOTE_ACTION_CALL = 21
@@ -690,8 +701,8 @@ class _RemoteObjectPipe(RemoteObjectPipe):
     def get_remote_obj_by_id(self, remote_id):
         return self._ros.get_remote_obj_by_id(remote_id)
 
-    def expose(self, name, obj, remote=True, methods=None, fields=None):
-        return self._ros.expose(name, obj, remote, methods, fields)
+    def expose(self, name, obj, remote=True, methods=None, fields=None, error=False):
+        return self._ros.expose(name, obj, remote, methods, fields, error)
 
     def hide(self, name):
         self._ros.hide(name)
@@ -918,11 +929,17 @@ class _RemoteObjectPipe(RemoteObjectPipe):
             if action_type == ROP_GET_EXPOSED:
                 exposed_name = data[1]
                 exposed = self._ros.get_exposed_by_name(exposed_name)
-                self._send_obj((ROP_GET_EXPOSED_RESULT, exposed))
+                if isinstance(exposed, ExposedObjectError):
+                    self._send_obj((ROP_GET_EXPOSED_ERROR, exposed.cause))
+                else:
+                    self._send_obj((ROP_GET_EXPOSED_RESULT, exposed))
                 continue
 
             if action_type == ROP_GET_EXPOSED_RESULT:
                 return data[1]
+
+            if action_type == ROP_GET_EXPOSED_ERROR:
+                raise data[1]
 
             if action_type == ROP_GET_PROXY_DEF:
                 remote_id = data[1]

